### ISSUE TITLE
Generic Filter expressions for NUnit tree

### ISSUE THREAD

**Issue by CharliePoole** [TestCentric/TestCentricRunner#1150]
So far we have discussed specific filters. But NUnit itself supports a general filter using the --where option of the console runner. The framework knows how to deal with the filters created by this option - (I wrote the parser)

Do we want this sort of functionality for the GUI? A command-line option does not seem very useful. Howe would we expose it the feature?

Decisions:

* Decide if we want this
* Decide how the user would activate it in the GUI
* Decide how it would display in the tree
* Decide what happens to the run commands if a filter is active.

**Comment by rowo360** [comment]
Here is an idea of how we could support the general NUnit filter:
First of all, I think that the existing filter functionality already supports a wide range of use cases. So I hope and expect that the majority of users won't miss anything. 
But nonetheless the NUnit filter supports more use cases and some power users will notice the gap. So we should try to support them also. But my idea is that we don't create a new UI to configure this filter - that might be too tough to add UI elements to configure all those operators and expressions. Instead we can use the existing text filtering textbox for this purpose.
My consideration is, if almost only power users use this, then they also know the syntax of the NUnit filter.

Here's a screenshot how this might look like:
<img src="https://github.com/user-attachments/assets/1972773c-1798-4ce8-bccd-75aee10f769f" width="400">

For example we might add a dropdown box at the end of the filter textbox. It can be used to select the filter mode: Text or NUnit. And of course we can guide the user with some Tooltips. But I hope that it's self-explanatory.

I imagine that the filtering itself can be performed by a Runner.Explore(testfilter) call. This call will return a XmlNode/TestNode which can be used to get a list of IDs which are matching the filter.
This idea would fit very well to the existing filtering (regarding display in tree and also to the run command when filter is active).

One drawback of this idea: the user can edit incorrect syntax of the filter statement. We must deal with this incorrect syntax somehow, provide some feedback and decide how to filter the tree in this case...

**Comment by rowo360** [comment]
@CharliePoole :
May I draw your attention back to this issue and my proposed solution. The core of the idea is that we would only support generic filters textually and don't create a separate UI for all those operators and options - see description above.
And another big advantage would also be that I think the effort for this would be moderate.

**Comment by CharliePoole** [comment]
@rowo360  Sorry I didn't respond earlier. 

I guess we are agreeing on my first "Decision point" - that we ought to support this.

Regarding who uses... the console only supports two options for deciding what should be run: --test and --where. So, for selecting anything except the name of the test, any user has to learn this syntax, which is pretty simple.

So, for the second decision, how to activate it, I agree that we simply need a text box. Using one text box for two things is a little confusing to me. What do the choices in the dropdown mean? I take it that "NUnit" means a "where clause" but what is "Text"?

I'm also uncertain about how you would combine filters created by various means into a single filter. Can you expand on that?

**Comment by rowo360** [comment]
I only had the --where clause in mind, but I think (if required) the idea can be extended to support also the --test clause. I'll use the term 'NUnit filter' for these.
The initial thought was that we support this NUnit filter only textually and all subsequent considerations are motivated by this. And since there exists already a textbox for filters, it was obvious to come to the conclusion to use it here too. However it's of course challenging to use one single textbox for slightly different purpose. How does the user get aware of it? 
Therefore, I added this dropdown box next to the textbox with the two options 'Text' and 'NUnit': either the text input is used as textual filter on any part of the tree ('Text') or the text input is used as a NUnit filter (--where) ('NUnit'). The terms for these options might be improved of course. But basicially the user has the opportunity to select between these two filter modes on his own.
If we find some good icons, we can replace the terms with these icons - getting a more compact representation.
I also believe that we need some tooltips to guide the user to the two options.

A somewhat more courageous approach would be to work without the drop-down box. The user simply enters something and it just works. If he starts his input with '--where' it's interpreted as a 'NUnit filter' otherwise it's a textual filter. I definitely like the simplicity of this approach. And I also propose for this approach that there's a tooltip guiding the user that he can simply enter text or a nunit filter.
Hmm, this approach will only fail if a user named a test/class/namespace '--where' and he wants to filter for it textual. However that will be an exceptional case.

A completely different idea regarding the UI:
Instead of combining the UI of the existing filtering and the NUnit filtering, we can separate them. So, currently there's a toolbox item to show/hide the filtering UI elements. We can replace it with a dropdown bow, so that the user can select the kind of filtering: NUnit or 'Simple'. Sorry, I don't come up with a good term for the existing filtering - but we need to define the terms.
In case of NUnit mode, there'll be only a textBox input. Here's a fake screenshot:

<img src="https://github.com/user-attachments/assets/38a836c5-f663-4e73-ad01-403424a2b02f" width="400">

So that's all regarding the UI - with respect to the simplicity I like the 2nd approach.

**Comment by rowo360** [comment]
Currently each filter (outcome, category, text) checks if the filter condition matches to a TestNode. According to the result the 'IsVisible' property of a TestNode is set. This property is used to decide if TestNode should be shown in the tree and if a TestNode should be included in a test run. 
If multiple filters are active all filter conditions must match to set the IsVisible flag - or in other words an AND operation is applied to the indivudal filter results to get a combined filter.

My consideration regarding the NUnit filter was that it can work identically. We can use the method Runner.Explore(nunitFilter) to get a list of Ids and so the NUnit filter can filter TestNodes based on their ID. And with that the IsVisible flag will be set accordingly.
And as soon as the IsVisible property is set, the Tree will show the correct nodes and a test run will respect only the visible nodes.

And I expect that it will work right away when combining filters: for example outcome and NUnit filter. Both filters will first check their filter condition on a TestNode, and if both conditions are fullfilled the IsVisible property of  the TestNode is set to true.
Combining outcome and NUnit filter provides some benefit, which is not possible with plain NUnit filter.
However when combining category filter and NUnit filter, the user might configure conflicting conditions. But that's up to the user, I don't see a general problem here.

**Comment by CharliePoole** [comment]
OK, I understand now. So let's back away for a moment from the UI, since there are so many different ways to do it, and talk about what I was hoping to accomplish: letting the framework do the filtering. 

1. I'd like to send NUnit (framework) a single filter that picks up everything selected, by any criterion.
2. I'd like to avoid use of the Id filter unless it's needed, i.e. grouping by duration or outcome.

I don't want the user to type in --test. I mentioned that because combining two options into one filter is something that is already done in the console runner where --test and --where become a single filter, which is then sent to NUnit. So, we already know how to create a single filter, usiing the AndFilter and OrFilter and NotFilter to combine them.

So my ideal flow would be this..

1. The UI allows the user to specify simple or advanced filtering in various ways.
2. We use that input to create a single filter object.
3. We send that filter object to the framework for exploring or running tests
4. We also use that filter object to decide which tree nodes should be displayed.

Clearly this is not a simple thing, that's why I made it a design issue, which we don't generally use. (BTW I mean software design here not UX design exclusively.) I imagine our "Filter Object" would not be as simple as an NUnit filter, although it should produce the same XML representation to be sent to NUnit. An NUnit filter doesn't know anything about Windows Trees and ours would need to know how to apply itself to a tree. This seems doable to me.

Back to the UI... I don't think a dropdown is a good answer. I agree that a where clause with NUnit's Test Selection Language is something that won't be used so often. I imagine something like a dialog box where you enter all the information. The existing filtering should be kept simple and use the existing toolbar.

Without getting down into the implementation details, can you see us doing this? It's too much for one PR, so we should figure out a sequence of different changes to accomplish it. One way I can see would be to create an advanced dialog and figure out how the user would bring it up. Then working on the basic filtering UI and populating the advanced dialog could proceed in parallel.

So... If there were an advanced filtering dialog that did "everything" including a where clause, what would we want to leave on the toolbar? 

If we can answer that last question clearly, then I think the "advanced" part can be postponed while we do some other stuff that is higher priority. But the answer is needed so we know what NOT to work on as part of "simple filtering".

**Comment by rowo360** [comment]
Before I start to try to answer, I will formulate the vision in my own words. Just to confirm that I got everything properly:
I will use the terms simple filtering for the existing filtering (outcome, category and text) and advanced filtering for the NUnit filtering ([NUnit's Test Selection Language](https://docs.nunit.org/articles/nunit/running-tests/Test-Selection-Language.html))

- The existing simple filtering is kept untouched.
- Untouched in the sense of the UI and how filtering is performed
For example: the category filtering and textual filtering is not transformed into a NUnit filter
- The existing simple filtering and the advanced filtering are kept separate.
- Separate in the sense that it is only possible to use either one or the other filtering. It's not possible to combine both approaches in any way.
- If advanced filtering is used, the actual filtering is done by the framework
- For that purpose we'll transform the user input into a NUnit filter object
- This NUnit filter object is used to explore/run tests, but also to decide which nodes will be displayed

**Comment by CharliePoole** [comment]
Good idea. I see parts where I did not explain fully, so I'll enlarge.

> The existing simple filtering is kept untouched.

Yes

> Untouched in the sense of the UI and how filtering is performed
For example: the category filtering and textual filtering is not transformed into a NUnit filter

No. Preferably, everything is transformed into a single NUnit filter. This is true now, we use an ID filter, which is an NUnit filter, but we don't create a _single_ filter combining all factors.

>The existing simple filtering and the advanced filtering are kept separate.
Separate in the sense that it is only possible to use either one or the other filtering. It's not possible to combine both approaches in any way.

Separate in the UI but not separate in what is sent to NUnit. However, we might want to keep them fully separate initially, until the advanced filtering is more fully developed.

> If advanced filtering is used, the actual filtering is done by the framework

Also, simple filtering. We always send a filter to the framework. Some runners do a separate run command per test, but we have never done that, we have always sent a single run command, which takes a filter as an argument.

> For that purpose we'll transform the user input into a NUnit filter object
This NUnit filter object is used to explore/run tests, but also to decide which nodes will be displayed


So...

```
 Simple Filter GUI  }                     { Apply to Tree
                    }---> GUI FILTER ---> {
AdvancedFilterGui   }                     { Create XML ---> Send to NUnit
```

Also, although I didn't say it earlier, I'm assuming that an Advanced Filter Dialog would include all the same features as are available in the Simple Filter GUI plus the use of a where clause.

Of course, this is all very high level and may discover things that cause it to change when we start working on the actual code.

**Comment by CharliePoole** [comment]
**Category filtering:**
In the past, for a simple category, I just hard-coded the necessary XML, e.g. `<filter><cat=CATEGORY/></filter>` (from memory, so may be wrong). I have not seen what you are doing now but if you are sending an id filter with an entry for every test, that can be a lot of traffic. For complex filters, there is an engine service that parses a where clause, btw.

For applying to the tree, I can easily write something that matches what I wrote for the framework. For now, you should keep doing what you are now doing rather than using explore, which creates another round trip with a lot of data.

**Textual filtering:**
You are correct that NUnit doesn't handle regular expressions on the test name. We should decide whether we need regex or if NUnit's filter with use of the `~-` operator is sufficient. IIRC, it is taken as equivalent to `StartsWith`.  Is there a strong enough need for regex that we should ask NUnit to provide a filter for it? Or should we just do less than you are now doing?
One option is to identify whether the text is a regular expression and process them differently, i.e. exact match vs starting string vs regex.

**Outcome filtering;**
We already use an nunit filter.... this use case, plus Duration, is why I created the id filter. At least my memory is that is what ou are using, if that means anything at all these days. :-) But since it can create a very large filter expression, it's best to avoid it when we can use a more terse expression.

BTW, an approach to consider would be take a general filter and create an id filter from it for applying to the tree. The code would be about the same but we might consider persisting that id filter until anything changed. Alternatively, the flags in each tree node will serve to persist this information as well. That's an implementation detail however.

-----

In all of this, I'd like to handle two tasks myself:

1. Creating a general filter that can be applied to the tree to replace the current approach.
2. Creating the advanced filtering dialog.

Both of these will be longish-running branches, which I'll periodically update from main. I'll ask you to do some reviewing of commits as I go.

In other news... I've recently had some time to work on this project but I'll be slowing down as this month progresses. I'm giving a new genealogy presentation in about three weeks, so I'll soon start spending more time to be ready for that. I'll review PRs you ask me to review by assigning me as the reviewer, so you can decide whether you need my review or not on a case by case basis.
